// @splitsoftware/browser-rum-agent/cjs/utils/context.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.languageVersion = exports.userAgent = exports.getRoute = exports.getUrl = exports.getConnectionType = exports.getUserAgent = void 0;
/* eslint-disable compat/compat */
var constants_1 = require("./constants");
/**
 * Get user agent if it is available
 */
function getUserAgent() {
    return typeof navigator === 'object' ? navigator.userAgent : constants_1.UNAVAILABLE;
}
exports.getUserAgent = getUserAgent;
/**
 * Get information about the system's connections ('wifi', 'cellular', etc)
 * More Information: https://developer.mozilla.org/en-US/docs/Web/API/Network_Information_API
 * */
function getConnectionType() {
    // @ts-expect-error `navigator.connection` property is not defined in lib.dom.d.ts but might exist
    var connection = typeof navigator === 'object' && (navigator.connection || navigator.mozConnection || navigator.webkitConnection);
    return connection ? connection.effectiveType : constants_1.UNAVAILABLE;
}
exports.getConnectionType = getConnectionType;
/**
 * Get the current page URL
 */
function getUrl() {
    return location.href; // equivalent to `document.URL`
}
exports.getUrl = getUrl;
/**
 * Get the current page URL without origin. i.e, path, query params and fragment, without schema, hostname and port
 */
function getRoute() {
    return location.pathname + location.search + location.hash;
}
exports.getRoute = getRoute;
exports.userAgent = getUserAgent();
exports.languageVersion = 'jsrum-' + '0.4.1';
